openverse_catalog/dags/maintenance/rotate_db_snapshots.py

```diff
@@ -0,0 +1,91 @@
+"""
+Manages weekly database snapshots. RDS does not support weekly snapshots
+schedules on its own, so we need a DAG to manage this for us.
+
+It runs on Saturdays at 00:00 UTC in order to happen before the data refresh.
+
+The DAG will automatically delete the oldest snapshots when more snaphots
+exist than it is configured to retain.
+
+Requires two variables:
+
+`AIRFLOW_RDS_ARN`: The ARN of the RDS DB instance that needs snapshots.
+`AIRFLOW_RDS_SNAPSHOTS_TO_RETAIN`: How many historical snapshots to retain.
+"""
+
+import logging
+from datetime import datetime
+
+import boto3
+from airflow.decorators import dag, task
+from airflow.models import Variable
+from airflow.providers.amazon.aws.operators.rds import RdsCreateDbSnapshotOperator
+from airflow.providers.amazon.aws.sensors.rds import RdsSnapshotExistenceSensor
+
+
+logger = logging.getLogger(__name__)
+
+DAG_ID = "check_silenced_dags"
+MAX_ACTIVE = 1
+
+AIRFLOW_RDS_ARN = Variable.get("AIRFLOW_RDS_ARN")
+AIRFLOW_RDS_SNAPSHOTS_TO_RETAIN = Variable.get(
+    "AIRFLOW_RDS_SNAPSHOTS_TO_RETAIN", default_var="7"
+)
```

Comment:
FWIW, it's generally a good idea (if we can help it) to refrain from calling `Variable.get` at the module level since that will issue a database call for each `.get` during each DAG parse interval. Since these values aren't needed until runtime, we can use some indirection Airflow provides to prevent having to assign them here. For example, many of the RDS operators have `db_identifier` in their [templated fields](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/rds/index.html#airflow.providers.amazon.aws.operators.rds.RdsCreateDbSnapshotOperator.template_fields). That means that we can use [the template variable reference](https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html#airflow-variables-in-templates) within a string for that value. Below is a diff which shows how this could work:

```diff
diff --git a/openverse_catalog/dags/maintenance/rotate_db_snapshots.py b/openverse_catalog/dags/maintenance/rotate_db_snapshots.py
index b24504c..ea13fc7 100644
--- a/openverse_catalog/dags/maintenance/rotate_db_snapshots.py
+++ b/openverse_catalog/dags/maintenance/rotate_db_snapshots.py
@@ -28,25 +28,18 @@ logger = logging.getLogger(__name__)
 DAG_ID = "rotate_db_snapshots"
 MAX_ACTIVE = 1
 
-AIRFLOW_RDS_ARN = Variable.get("AIRFLOW_RDS_ARN")
-AIRFLOW_RDS_SNAPSHOTS_TO_RETAIN = Variable.get(
-    "AIRFLOW_RDS_SNAPSHOTS_TO_RETAIN", default_var="7"
-)
-
 
 @task()
-def delete_previous_snapshots():
+def delete_previous_snapshots(rds_arn: str, snapshots_to_retain: int):
     rds = boto3.client("rds")
     # Snapshot object documentation:
     # https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_DBSnapshot.html
     snapshots = rds.describe_db_snapshots(
-        DBInstanceIdentifier=AIRFLOW_RDS_ARN,
+        DBInstanceIdentifier=rds_arn,
     )
 
-    to_retain = int(AIRFLOW_RDS_SNAPSHOTS_TO_RETAIN)
-
-    if len(snapshots) <= to_retain or not (
-        snapshots_to_delete := snapshots[to_retain:]
+    if len(snapshots) <= snapshots_to_retain or not (
+        snapshots_to_delete := snapshots[snapshots_to_retain:]
     ):
         logger.info("No snapshots to delete.")
         return
@@ -68,13 +61,16 @@ def delete_previous_snapshots():
     catchup=False,
     # Use the docstring at the top of the file as md docs in the UI
     doc_md=__doc__,
+    # Render JSON deserialized values as native objects rather than strings
+    render_template_as_native_obj=True,
 )
 def rotate_db_snapshots():
     snapshot_id = "airflow-{{ ds }}"
+    db_identifier = "{{ var.value.AIRFLOW_RDS_ARN }}"
     create_db_snapshot = RdsCreateDbSnapshotOperator(
         task_id="create_snapshot",
         db_type="instance",
...skipping...
     rds = boto3.client("rds")
     # Snapshot object documentation:
     # https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_DBSnapshot.html
     snapshots = rds.describe_db_snapshots(
-        DBInstanceIdentifier=AIRFLOW_RDS_ARN,
+        DBInstanceIdentifier=rds_arn,
     )
 
-    to_retain = int(AIRFLOW_RDS_SNAPSHOTS_TO_RETAIN)
-
-    if len(snapshots) <= to_retain or not (
-        snapshots_to_delete := snapshots[to_retain:]
+    if len(snapshots) <= snapshots_to_retain or not (
+        snapshots_to_delete := snapshots[snapshots_to_retain:]
     ):
         logger.info("No snapshots to delete.")
         return
@@ -68,13 +61,16 @@ def delete_previous_snapshots():
     catchup=False,
     # Use the docstring at the top of the file as md docs in the UI
     doc_md=__doc__,
+    # Render JSON deserialized values as native objects rather than strings
+    render_template_as_native_obj=True,
 )
 def rotate_db_snapshots():
     snapshot_id = "airflow-{{ ds }}"
+    db_identifier = "{{ var.value.AIRFLOW_RDS_ARN }}"
     create_db_snapshot = RdsCreateDbSnapshotOperator(
         task_id="create_snapshot",
         db_type="instance",
-        db_identifier=AIRFLOW_RDS_ARN,
+        db_identifier=db_identifier,
         db_snapshot_identifier=snapshot_id,
     )
     wait_for_snapshot_availability = RdsSnapshotExistenceSensor(
@@ -85,7 +81,10 @@ def rotate_db_snapshots():
         target_statuses=["available"],
     )
 
-    create_db_snapshot >> wait_for_snapshot_availability >> delete_previous_snapshots()
+    create_db_snapshot >> wait_for_snapshot_availability >> delete_previous_snapshots(
+        rds_arn=db_identifier,
+        snapshots_to_retain="{{ var.json.AIRFLOW_RDS_SNAPSHOTS_TO_RETAIN }}",
+    )
 
 
 rotate_db_snapshots()
```

Reply:
I would not say this DAG is an edge case and yes we should change those other instances if possible to! We have this issue for tracking the removal of the API key retrieval at the module level: https://github.com/WordPress/openverse-catalog/issues/688

I'll make another for changing the other instances 🙂

Reply:
Great. This seems like something perfect for https://github.com/hchasestevens/bellybutton along with using datetime instead of pendulum.